src/lazy/binary/immutable_buffer.rs

```diff
@@ -157,6 +159,257 @@ impl<'a> ImmutableBuffer<'a> {
+        let unsigned_value = flex_uint.value();
+
+        // If the encoded FlexInt required `N` bytes to encode where `N` is fewer than 8, then its
+        // u64 value will have `8 - N` leading zero bytes. If the highest bit in the encoding was a
+        // 1, then then number is negative and we need to flip all of those leading zeros to ones.
+        // Look at the original input to see if the highest bit was a zero (positive) or one (negative).
+        let last_explicit_byte = self.bytes()[flex_uint.size_in_bytes() - 1];
+        let sign_bit = 0b1000_0000 & last_explicit_byte;
+        let signed_value = if sign_bit == 0 {
+            unsigned_value as i64
+        } else {
+            // Flip all of the leading zeros in the unsigned value to ones and then re-interpret it
+            // as a signed value.
+            let mask = ((1u64 << 63) as i64) >> unsigned_value.leading_zeros();
+            (unsigned_value as i64) | mask
+        };
+
+        Ok((
+            FlexInt::new(flex_uint.size_in_bytes(), signed_value),
+            remaining,
+        ))
+    }
+
+    /// Reads a [`FlexUInt`] from the buffer.
+    #[inline]
+    pub fn read_flex_uint(self) -> ParseResult<'a, FlexUInt> {
+        const COMMON_CASE_INPUT_BYTES_NEEDED: usize = 8;
+
+        // We want to minimize the number of branches that happen in the common case. To do this,
+        // we perform a single length check, making sure that the buffer contains enough data to
+        // represent a FlexUInt whose continuation bits fit in a single byte (i.e. one with 7 or
+        // fewer bytes of magnitude). If the buffer doesn't have at least 8 bytes in it or the
+        // FlexUInt we find requires more than 8 bytes to represent, we'll fall back to the general
+        // case.
+        if self.len() < COMMON_CASE_INPUT_BYTES_NEEDED || self.bytes()[0] == 0 {
+            // `read_flex_uint_slow` is marked #[cold] to discourage inlining it, which keeps
+            // this method small enough that the code for the common case can be inlined.
+            return self.read_flex_uint_slow("reading a FlexUInt");
+        }
+
+        let flex_uint = Self::read_small_flex_uint(self.bytes());
+        let remaining = self.consume(flex_uint.size_in_bytes());
+        Ok((flex_uint, remaining))
+    }
+
+    /// Helper method that reads a [`FlexUInt`] with 7 or fewer bytes of magnitude from the buffer.
+    // Caller must confirm that `bytes` has at least 8 bytes.
+    #[inline]
+    fn read_small_flex_uint(bytes: &[u8]) -> FlexUInt {
+        debug_assert!(bytes.len() >= 8);
+        let num_encoded_bytes = bytes[0].trailing_zeros() as usize + 1;
+        let num_encoded_bits = 8 * num_encoded_bytes;
+        // Get a mask with the low 'n' bits set
+        // TODO: Should this be a const cache of num_encoded_bits -> mask?
+        let mask = 1u64
+            .checked_shl(num_encoded_bits as u32)
+            .map(|v| v - 1)
+            .unwrap_or(u64::MAX);
```

Comment:
🗺️ I think making a `const` cache for this would get us a modest speedup, but the performance is already much better than `VarUInt`.